### PR TITLE
Add query and filter context documentation

### LINKED ISSUE #3068: Query context

### PR DESCRIPTION
**Pull request by kolchfa-aws**
Fixes #3068 

### Checklist
- [x ] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: _opensearch/query-dsl/query-filter-context.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+---
+layout: default
+title: Query and filter context
+parent: Query DSL
+nav_order: 5
+---
+
+# Query and filter context
+
+Queries consist of query clauses, which can be run in a [_filter context_](#filter-context) or [_query context_](#query-context). A query clause in a filter context asks the question "_Does_ the document match the query clause?" and returns matching documents. A query clause in a query context asks the question "_How well_ does the document match the query clause?" and returns matching documents, providing the relevance of each document in the [_relevance score_](#relevance-score).
```
Comment (cwillum) [inline _opensearch/query-dsl/query-filter-context.md:10]: RE: "A query clause in a filter context asks the question "_Does_ the document match the query clause?" and returns matching documents."
Does a query clause return documents regardless of the answer? In other words, will it return documents even when the answer is "No"?

Suggestion only:
A query clause in a query context asks the question "_How well_ does the document match the query clause?", returns matching documents, and provides the relevance of each document in the [_relevance score_](#relevance-score). (or some variation on that.)

File: _opensearch/query-dsl/query-filter-context.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+## Relevance score
+
+A _relevance score_ measures how well a document matches a query. It is a positive floating-point number that OpenSearch records in the `_score` metadata field for each document. A higher score corresponds to a more relevant document. While different query types calculate relevance scores differently, all query types take into account whether a query clause it is run in a filter or query context.
```
Comment (cwillum) [inline _opensearch/query-dsl/query-filter-context.md:14]: "...  all query types take into account whether a query clause *it is run in* a filter or query context."
meaning here?

... whether a query clause runs in a filter or query context. ?

File: _opensearch/query-dsl/query-filter-context.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+# Query and filter context
+
+Queries consist of query clauses, which can be run in a [_filter context_](#filter-context) or [_query context_](#query-context). A query clause in a filter context asks the question "_Does_ the document match the query clause?" and returns matching documents. A query clause in a query context asks the question "_How well_ does the document match the query clause?" and returns matching documents, providing the relevance of each document in the [_relevance score_](#relevance-score).
+
+## Relevance score
+
+A _relevance score_ measures how well a document matches a query. It is a positive floating-point number that OpenSearch records in the `_score` metadata field for each document. A higher score corresponds to a more relevant document. While different query types calculate relevance scores differently, all query types take into account whether a query clause it is run in a filter or query context. 
+
+Use query clauses that should affect the relevance score in a query context, and all other query clauses in a filter context.
```
Comment (cwillum) [inline _opensearch/query-dsl/query-filter-context.md:16]: Use query clauses that have an effect on the relevance score in a query context, and use all other query clauses in a filter context.

Does this preserve the meaning?

**Review by cwillum — APPROVED** [review]
LGTM

File: _opensearch/query-dsl/query-filter-context.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+To improve performance, OpenSearch caches frequently used filters.
+
+## Query context
+
+A query clause in a query context asks the question "_How well_ does the document match the query clause?", which does not have a binary answer. A query context is suitable for a full-text search, where you not only want to receive matching documents, but also to determine how relevant each document is. For example, if you have an index with the complete works of Shakespeare, you might use a query context for the following searches:
+
+- Find documents that contain the word `dream`, including its various forms (`dreaming` or `dreams`) and synonyms (`contemplate`).
+- Find documents that match the words `long live king`.
+
+For a query context, every matching document contains a relevance score in the `_score` field, which you can use to [sort]({{site.url}}{{site.baseurl}}/opensearch/search/sort) documents by relevance.
```
Comment (Naarcha-AWS) [inline _opensearch/query-dsl/query-filter-context.md:53]: Does this link work without the trailing slash?

```suggestion
For a query context, every matching document contains a relevance score in the `_score` field, which you can use to [sort]({{site.url}}{{site.baseurl}}/opensearch/search/sort/) documents by relevance. 
```

File: _opensearch/query-dsl/query-filter-context.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+{: .tip}
+
+## Filter context
+
+A query clause in a filter context asks the question "_Does_ the document match the query clause?", which has a binary answer. For example, if you have an index with student data, you might use a filter context to answer the following questions:
+
+- Is the student's `honors` status set to `true`?
+- Is the student's `graduation_year` in the 2020--2022 range?
+
+For a filter context, OpenSearch simply returns matching documents without calculating a relevance score. Thus, you should use a filter context for fields with exact values.
```
Comment (Naarcha-AWS) [inline _opensearch/query-dsl/query-filter-context.md:26]:
```suggestion
For a filter context, OpenSearch returns matching documents without calculating a relevance score. Thus, you should use a filter context for fields with exact values. 
```

File: _opensearch/query-dsl/query-filter-context.md (hunk 2 above)
Comment (Naarcha-AWS) [reply _opensearch/query-dsl/query-filter-context.md:14]: Do we have an example of the `_score` metadata field inside a document? Maybe a link somewhere else in the documentation?

**Review by Naarcha-AWS — APPROVED** [review]
A couple of small suggestions, otherwise LGTM

File: _opensearch/query-dsl/query-filter-context.md (hunk 1 above)
Comment (kolchfa-aws) [reply _opensearch/query-dsl/query-filter-context.md:10]: No, in a filter context, if the document does not match, it is not returned.

File: _opensearch/query-dsl/query-filter-context.md (hunk 3 above)
Comment (kolchfa-aws) [reply _opensearch/query-dsl/query-filter-context.md:16]: I don't think it's quite the same. Whether or not a query clause affects the relevance score is not intrinsic to a query clause. If you want it to affect the relevance score, you should put it in a query context, and if you don't, then you put it in a filter context.

File: _opensearch/query-dsl/query-filter-context.md (hunk 4 above)
Comment (kolchfa-aws) [reply _opensearch/query-dsl/query-filter-context.md:53]: Yes, it works without a trailing slash. A trailing slash is optional.

File: _opensearch/query-dsl/query-filter-context.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
+---
+layout: default
+title: Query and filter context
+parent: Query DSL
+nav_order: 5
+---
+
+# Query and filter context
+
+Queries consist of query clauses, which can be run in a [_filter context_](#filter-context) or [_query context_](#query-context). A query clause in a filter context asks the question "_Does_ the document match the query clause?" and returns matching documents. A query clause in a query context asks the question "_How well_ does the document match the query clause?", returns matching documents, and provides the relevance of each document in the [_relevance score_](#relevance-score).
```
Comment (natebower) [inline _opensearch/query-dsl/query-filter-context.md:10]:
```suggestion
Queries consist of query clauses, which can be run in a [_filter context_](#filter-context) or [_query context_](#query-context). A query clause in a filter context asks the question "_Does_ the document match the query clause?" and returns matching documents. A query clause in a query context asks the question "_How well_ does the document match the query clause?", returns matching documents, and provides the relevance of each document in the form of a [_relevance score_](#relevance-score).
```

File: _opensearch/query-dsl/query-filter-context.md (hunk 3 above)
Comment (natebower) [reply _opensearch/query-dsl/query-filter-context.md:37]: Do we mean something like "Use query clauses that you want to affect the relevance score in a query context, and use all other query clauses in a filter context"?

**Review by natebower — COMMENTED** [review]
@kolchfa-aws Please see my comments and changes and let me know if you have any questions. Thanks!

**Pull request merged**